Records with unboxed fields

eliberis opened the pull request:
This pull request implements unboxing of record fields that are themselves tuples or records. 
## Motivation: avoiding an extra indirection when accessing record fields

Consider the following example:

``` OCaml
type point = { x : int; y : int }
type triangle = { a : point; b : point; c : point }
```

Currently an instance of `triangle` would be represented at runtime as a block of 4 words: 1 header word and 3 pointers to instances of `point`). So, to access individual `point` data, the runtime will have to follow one of these pointers. While it's not a big problem in this example, this indirection overhead can hurt the performance if you have big, complex, and deeply nested records which are accessed often. 

A sensible optimisation here would be to 'unpack' nested record fields into the parent record, that is fields of `point` into the `triangle` record, like this:

``` OCaml
type triangle = { ax : int; ay : int; bx : int; by : int; cx : int; cy : int }
```

But this loses the abstraction we had by having a `point` type. 

This memory layout optimisation could be done at compile time if the programmer asks the compiler to perform unboxing of some record fields, while keeping the same "dot" syntax for accessing fields (`my_triangle.a.x`).
## Implemented language extension

To request a field to be unboxed, the programmer can annotate a field with the `[@unboxed]` attribute:

``` OCaml
type triangle = { a : point [@unboxed];
                  b : point [@unboxed];
                  c : point; (* Some fields can be left boxed *)
                }
```

This attribute was first introduced in [`GPR#{606}`](https://github.com/ocaml/ocaml/commit/d5a6e50ebee73ff98c4179bba7570cdd9e488a35), and because it is semantically similar, it was reused in this feature for consistency. 

As mentioned above, there are no other changes to the user-facing part of the language --- unboxed fields are accessed in the same way as the non-unboxed ones. 

Nesting of records with unboxed fields is supported. Unboxing can be requested in all types of records: regular records (`type t = { ... }`), "inline" records (`type t = T of { ... } | ...`)  and extension variants (`type t += T of { ... }`).
### Restrictions

A field cannot be unboxed if:
- If it's type is recursive or mutually recursive with the parent record type;
- If it has an optimised floating point representation (float array);
- If it has mutable fields. But the field itself can be mutable, that is:

``` OCaml
type triangle = { mutable a : point [@unboxed]; ... } (* is OK *)

type mut_point = { mutable x : int; ... }
type triangle' = { a : mut_point [@unboxed]; ...} (* produces a compile error *)
```
- If it's type is already `[@@unboxed]` (as per `GPR#{606}`).
## Implementation

This change was implemented at the Lambda language level (`lambda.mli`), mostly in `translcore.ml` which handles record creation, field access and update.
### Memory layout

When a label is marked as `[@unboxed]`, the parent record allocates enough space in its block to accommodate the field's block. Since we only allow unboxing of records and tuples, which have a tag 0, we don't need to record it in the parent block. An earlier example would look as follows at runtime:

```
block of a triangle instance with all 3 points unboxed:
┌─────┬───────┬───────┬───────┬───────┬───────┬───────┐
│ tag │   x   │   y   │   x   │   y   │   x   │   y   │
└─────┴───────┴───────┴───────┴───────┴───────┴───────┘

```

Note that this means that the physical size of a record block is no longer just a number of fields it has, because now one logical field can occupy more than 1 word. 
### Field projection

When a programmer wants to access an unboxed field (e.g. `my_triangle.a`), a new block with tag 0 is created, which is then populated by copying respective fields (words) from the parent record. This is a big overhead, but we rely on Flambda optimising out this allocation (see Benchmarking section below). This annotation be beneficial even if Flambda is not enabled: instead of creating lots of long-lived objects, the program would create very few large long-lived objects and only allocate short-lived projections on demand.
#### Note on `record_representation`

Prior to this pull request, despite being quite similar, regular, inline, extension, unboxed, float records were each given their own variant representation. With the introduction of unboxed fields, the number of representations would've increased almost twofold (`regular_with_unboxed_fields`, `inlined_with_unboxed_fields`, ...). To mitigate this, the `record_representation` was simplified to remove overlapping representations (see `types.mli`). 
## Benchmarking

It is difficult to show reduced access time on small toy examples, but we can show that Flambda successfully optimises extra allocations introduced by this change.

A micro-benchmark is available [here](https://github.com/eliberis/ocaml-unboxing-performance/blob/master/main.ml). We obtained the following results using an Flambda-enabled compiler:

```
┌────────────────────────────────────────────────────────┬──────────┬─────────┬────────────┐
│ Name                                                   │ Time/Run │ mWd/Run │ Percentage │
├────────────────────────────────────────────────────────┼──────────┼─────────┼────────────┤
│ Access boxed (depth = 1)                               │  20.16ns │  32.00w │     20.87% │
│ Access unboxed (depth = 1)                             │  11.50ns │  20.00w │     11.91% │
│ Access boxed (depth = 2)                               │  96.62ns │ 138.00w │    100.00% │
│ Access unboxed (depth = 2)                             │  39.84ns │  74.01w │     41.23% │
│ Access boxed (depth = 2), with-style record creation   │  64.89ns │  90.00w │     67.16% │
│ Access unboxed (depth = 2), with-style record creation │  44.44ns │  74.01w │     45.99% │
└────────────────────────────────────────────────────────┴──────────┴─────────┴────────────┘
```

A detailed explanation of these numbers is at the bottom of the benchmark code
(makes references to types defined in the code).  

Special thanks to @lpw25 for mentoring this project.

eliberis committed:
Support for records with unboxed fields.

Add new record representation that allows records to have child records and tuples unboxed into them. See PR description.

eliberis committed:
Documentation update

alainfrisch commented on bytecomp/translcore.ml:
```diff
@@ -1283,57 +1295,115 @@ and transl_record loc env fields repres opt_init_expr =
+              slot :: ll, Pgenval :: shape
+            | No_inline | Inlined -> ll, shape
+          in
+          if not has_unboxed_fields then
```
This conditional seems to be the reason why the `has_unboxed_fields` property is kept in the record_representation.  I think I would prefer to have the check be done here explicitly (it would be very cheap) instead of being stored in the Types representation (adding redundancy to it).

eliberis replied:
Yes, it's for this optimisation and it's also used a few times in `typing/includecore.ml`. 

It could be replaced by something like `List.exists (fun lbl -> lbl.lbl_unboxed) lbls`, but I thought it's useful to keep this property around instead of duplicating the code. If you feel strongly against having an extra field in the type, push, I'm happy to remove it --- I have no strong preference here. Thanks!

alainfrisch commented:
Thanks for the interesting contribution!

Do you think the implementation technique opens the door to (or at least does not complicate) adding more layout-tweaking control, such as allowing multiple "small integer" fields to be packed in a single word (e.g. packing multiple bools)  or unboxing sum types (keeping one field for the tag and reserving the maximum number of fields across all constructors)?

eliberis commented:
Thanks!

I don't think it will complicate packing multiple small fields into a single word --- it currently relies on `Pfield`/`Psetfield` to do field manipulation, so as long as those primitives are updated to support 'packed' fields (or new similar ones are added), this feature could be changed to use them instead.

eliberis commented:
Tests seem to fail because camlp4 fails to compile. I have a working version of camlp4 [here](https://github.com/eliberis/camlp4), but I'll wait for this discussion to settle before making a pull request.

gasche commented:
When we discussed inline records, Arthur Charguéraud was of the opinion that we should think about a [more general feature](http://caml.inria.fr/mantis/view.php?id=5528#c10510) that would allow the inclusion/embedding of arbitrary product types inside a block. The idea involved keeping the header word side, and play clever with the runtime to allow referring to the block inline instead (I'm not sure if it actually works, but if so it may lift the limitation on mutable fields).

I think the current proposal looks nice, but that it might also be time to consider whether a more general approach such as the one proposed by Arthur wouldn't be a more future-proof step forward. Or, maybe, whether we can make sure that any specific representation choice we make today (for example, in terms of the API exposed at the C level to access those inline fields) is compatible with a different, more generation representation we could decide to adopt in the future. With the currently proposed implementation, would anything break if tomorrow we modified the compiler to include the header word as well?

lpw25 commented:
I'm not sure whether a general approach based on something like `Infix_tag` would work very well, and even if it did I would expect it to be available in addition to the proposal in this PR since the approaches have different pros and cons.

But more importantly for the purposes of this PR:

> With the currently proposed implementation, would anything break if tomorrow we modified the compiler to include the header word as well?

No, I think it is future-proof against that kind of change.

eliberis committed:
Update bootstrap executables

gasche commented:
Do I understand correctly (I haven't looked at the code yet) that this PR includes no support for accessing the unboxed fields from the C side? Does that mean, in practice, that people would just not access those fields from the C side (I feel somehow skeptical that this discipline would hold), or compute field positions using their knowledge of the current representation? This would endanger the representation-independence claim. Would it be possible to include pointer-computing macros for access within an unboxed field to also make C-side access future-proof?

(I'm not sure there is design space for two slightly different inlining strategies.)

eliberis commented:
Yes, you're right, this PR doesn't exclude any explicit support for the C side, only the manual update. I'll look into making the C integration a bit more friendly.

bluddy commented:
@gasche I think there is potential for different inlining strategies. We must assume the programmer is smart enough to know what he is doing, and the potential performance impact of each choice.

For example, the next step from this PR IMO is to have unboxed float and int64 fields. This would require modifying the record representation to place pointers separate from non-pointers, and to add an index field pointing to the partition point, which is precisely haskell's object representation. However, the cost of the extra field, shuffling of order, and resulting backwards incompatibility means that the programmer must actively choose this strategy -- a strategy which will co-exist with the tamer strategy proposed here.

eliberis commented:
@gasche Re C bindings: 

The field position information is kept as a part of the type information which is not present at runtime. So the user will have to recompute field positions themselves.

Still, I imagine some convenience functions could be:
- Field setter: setting an unboxed field of a record given a position and a value (block) to set it to. 
- Field getter / projection: copying a range of bytes into a new block.

Both would be essentially a memcpy with some additional checks and making sure that GC interaction works correctly.

Do you think it would be worth adding these?

gasche commented:
(I would like to apologize in advance about the fact that I'm not very familiar with C-interface code and may say or have already said dummy things about it.) Currently `Field(v, n)` is a l-value denoting the `n`-th field of the record value `v`. In a record that has an inlined sub-record at position `n1`, `Field(Field(v, n1), n2)` is not a correct way to access the `n2`-th field of the sub-record, as `Field(v, n1)` does not start with a valid block header. Maybe we could have a `Subfield` macro such that `Subfield(Field(v, n1), n2)` is correct?

As you suggest one would also need some kind of `Store_subfield` macro. I don't know about macros for setting and reading the whole subrecord in one go (implicitly allocating in the reading/projection case); implicitly allocating a new record feels a bit too magical to me, but indeed we would probably need an API that can express such an operation.

Another problem I don't have a nice solution for is how to access the fields of the original record that are after the inlined record. It seems natural to use `Field(v, n1 + inner_record_length + offset_from_n1)` with the proposed implementation, and I see no macro-tic way to do it that would both (1) port nicely to a representation with the header word and (2) be sufficiently clear and natural that programmers would prefer it over the direct computation.

bluddy commented:
We're leaving the domain of uniform data representation and I agree that there simply isn't a clean generic way to do this anymore. Probably the best approach at this point is to have the compiler/a tool generate C macros/consts with the offset of every field, and using that. Imagine a record with an unboxed field, a normal field, another unboxed field which is a record containing more unboxed fields, yet another normal field etc, and then trying to calculate offsets correctly. This simply doesn't scale without the compiler helping the C user out.

No need to hold up this PR with waiting for such a tool IMO.

eliberis commented:
@gasche I'd agree with @bluddy here that there probably isn't a nice generic way of doing this. In your suggestion, `Subfield` would effectively be an extra addition, which IMO is not that useful as a macro, here's why:

If the programmer requests unboxing, they would have to hardcode individual record field sizes that they've requested in OCaml-land (since the information erased after typechecking). So effectively they would already have to explicitly rely on a specific memory arrangement they've chosen, so they might as well maintain appropriate record field positions themselves (that properly take into account which fields are unboxed at which level of nesting)

I was told that the binding generator is long overdue, and IMO until then manual offsetting seems reasonable for people who actually want to have these new unboxed records available in C-land.

braibant commented:
This looks really nice. I have one minor concern: does this guarantee `t.a == t.a` (with and without flamba)?

gasche commented:
Well, note that

```
# type t = { a : float };;
# let t = { a = 0.2 };;
# t.a == t.a;;
- : bool = false
```

braibant commented:
Yes, float boxing is painful (but one should not use physical equality on floats). But I agree with your point which I guess is that it's already not the case for all variations of `t` and `a`s.

alainfrisch commented:
One should not use physical equality on immutable values, and this PR excludes unboxing mutable records.

alainfrisch commented:
I agree with @eliberis : the current C API is enough to work on unboxed records, as long the layout is specified; the programmer has to compute offsets anyway.  It might be useful to provide a C function to extract an unboxed record (i.e. something similar to caml_array_sub), but I doubt it would be common for C bindings to use this operation (as opposed perhaps to getting/setting fields under an unboxed record).

eliberis committed:
Disallow unboxing of private types

The purpose of private types is to disallow creating instances of a type marked as 'private'. However, now projecting an unboxed field implicitly creates a new block for that field, and if the field's type is 'private', this breaks the intuition behind private types. Such projection would break programs of users who (wrongly) rely on physical equality of instances of a private type, so this commit disallows unboxing of private types.

lpw25 commented:
We discussed this at the developer's meeting. The general feeling was that a clearer motivation was needed along with a description of how this `[@unboxed]` attribute will be extended to more cases in the future. I'll try and write these myself in the near future.

DemiMarie commented:
@bluddy I prefer a different approach, albeit one other than the one OCaml has historically used: quit accessing OCaml values from C code, and instead provide the needed OCaml %-primitives and tooling support to efficiently access C data types (structs, unions, etc) _from OCaml_.  That's how GHC (and all other Haskell implementations, via the standardized Foreign Function Interface) do it, and it has worked out extremely well for them.  This technique allows the user to write code that is 100% independent of OCaml's internals, and (with adequate tool support, analogous to hsc2hs or c2hs) is also robust to changes in C struct layout.

fpottier commented:
I was just pointed to this discussion by @mshinwell.
@charguer and I would very much like to have something like this, as we are interested in implementing one abstract data type on top of another abstract data type, without having to pay for an indirection. (Marketing people would call it "zero-cost abstraction".)

E.g., suppose you have an abstract type of sets that does not have a constant-time `cardinal` operation, and you would like to define a new abstract type of sets that have this operation. You would like to define the new type of sets as a pair of an underlying set and an integer field. But you would prefer not to allocate two records at runtime, and chase pointers.

I should point out that this PR does not seem to go quite as far as we would like, as we would like this optimization to work also in the case where the inner (unboxed) record is mutable, and is conceptually uniquely owned by the outer record. This situation seems more difficult to handle. We haven't thought about it in detail, but it seems that the compiler would have to keep track of this unique association between outer and inner records (e.g., "t.a holds a unique pointer to p"), and transform updates to the inner record accordingly (e.g., "p.x <- 0" should be transformed to "t.a.x <- 0").

gasche commented:
@lpw25 I wonder what is your current take on this work, for which we were unfortunately unable to take a decision. Is it worth having this PR open now that you are working on a more general low-level-representation story with @stedolan? Does the proposed design integrate nicely-enough with future work?

lpw25 commented:
I think we can close this PR now. It was nice work, but we plan to have a more general solution to the problem based on [the unboxed types proposal](https://github.com/ocaml/RFCs/blob/unboxed-types/rfcs/unboxed-types.md).
